### PR TITLE
replaced original BLE_iBeacon.ino example with a known-good one

### LINKED ISSUE #7360: Please replace BLE_iBeacon with a more fleshed-out example.
### Related area

iBeacon example

### Hardware specification

ESP32

### Is your feature request related to a problem?

Forum and github searches reveal that many people have problems getting ESP32 boards running https://github.com/espressif/arduino-esp32/blob/master/libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino to be detected by strict iBeacon scanners. Even when a scanner does detect an ESP32 iBeacon, many fields are missing. I came across a discussion [here](https://github.com/nkolban/esp32-snippets/issues/385#issuecomment-362535434) that links to reworked example code [here](https://pastebin.com/CNZLtLNK). This worked out of the box for me, whereas BLE_iBeacon.ino only worked partially on some scanners and not at all on others. (EDITED: updated pastebin link to a the full solution)

### Describe the solution you'd like

Either _replace_ the legacy BLE_iBeacon.ino with [31337Ghost](https://github.com/31337Ghost)'s reference code, or incorporate its changes into the legacy BLE_iBeacon.ino.

### Describe alternatives you've considered

I searched github and forums, and it seems as if most people end up having to modify BLE_iBeacon.ino to achieve basic functions. I don't think there's really an alternative, other than everyone performing the same forum searches and performing the same workarounds.

### Additional context

I recommend that devs try the legacy BLE_iBeacon code, then switch to the suggested new reference code, and compare the functionality in various scanners.

### I have checked existing list of Feature requests and the Contribution Guide

- [X] I confirm I have checked existing list of Feature requests and Contribution Guide.

### PR DESCRIPTION
**Pull request by Human**
## Description of Change
I had no luck getting the original `BLE_iBeacon.ino` code to work as expected with iBeacon scanners, but the included reference code worked completely. Thanks to @31337Ghost for writing it! I am simply submitting the PR.
## Tests scenarios
Tested on DOIT ESP32 DEVKIT V1, using various scanner client apps on Android devices.

## Related links
Closes #7360

### CODE REVIEW

**Comment by P-R-O-C-H-Y** [comment]
@Human Can you please take a look on my comment above? Thanks

**Comment by Human** [comment]
@P-R-O-C-H-Y It will take me some time to be able to test this, but which app on which platform did you use?

File: libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino [hunk 1]
```diff
@@ -1,104 +1,131 @@
...
-  pAdvertising->stop();
-  Serial.printf("enter deep sleep\n");
-  esp_deep_sleep(1000000LL * GPIO_DEEP_SLEEP_DURATION);
-  Serial.printf("in deep sleep\n");
-}
-
-void loop() {
-}
+/*
+   Based on 31337Ghost's reference code from https://github.com/nkolban/esp32-snippets/issues/385#issuecomment-362535434
```
Comment (VojtechBartoska) [inline libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino:2]: @Human Please, it seems a lot of lines are still similar to the example created by Neil Kolban. Can you please mentioned in the header that this example is "modified" version by "31337Ghost" of the original example from Neil? Thanks

**Comment by P-R-O-C-H-Y** [comment]
> @P-R-O-C-H-Y It will take me some time to be able to test this, but which app on which platform did you use?

@Human I did test on ESP32-C3 and with HONOR phone running Android 10 + app called nRF Connect (also tried another app, but got same behaviour).

File: libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino (hunk 1 above)
Comment (Human) [reply libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino:2]: Please check revised comment.

**Comment by Human** [comment]
@P-R-O-C-H-Y I tried your recommended app, and I do not get a `CONNECT` option for the ESP32 iBeacon running this code. I do see the option for a BlueCharm beacon and my TV (for some reason).

**Comment by Human** [comment]
@P-R-O-C-H-Y I was able to reproduce what you saw, but I don't know enough about BLE or iBeacons to fix the issue.

File: libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino [hunk 2]
```diff
@@ -1,104 +1,132 @@
...
+    void onConnect(BLEServer* pServer) {
+      deviceConnected = true;
+      Serial.println("deviceConnected = true");
+    };
+
+    void onDisconnect(BLEServer* pServer) {
+      deviceConnected = false;
+      Serial.println("deviceConnected = false");
+    }
+};
```
Comment (P-R-O-C-H-Y) [inline libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino:42]: Can you please start the advertising again after device is disconnected? I have tested this and now there is no need to restart the device. After disconnection, the ESP32 is visible and connectable again :)

```
class MyServerCallbacks: public BLEServerCallbacks {
     void onConnect(BLEServer* pServer) {
       deviceConnected = true;
       Serial.println("deviceConnected = true");
     };

     void onDisconnect(BLEServer* pServer) {
       deviceConnected = false;
       Serial.println("deviceConnected = false");
       
       // Restart advertising to be visible and connectable again
       BLEAdvertising* pAdvertising;
       pAdvertising = pServer->getAdvertising();
       pAdvertising->start();

     }
 };
```

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@Human please take a look on my comment :)

**Review by P-R-O-C-H-Y — APPROVED** [review]

File: libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino (hunk 2 above)
Comment (P-R-O-C-H-Y) [reply libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino]: @Human Thanks, please take a look on me-no-devs comment.

Comment (Human) [reply libraries/BLE/examples/BLE_iBeacon/BLE_iBeacon.ino]: @me-no-dev I pared it down to my three commits and force pushed. Please let me know if this is what you wanted.

**Pull request merged**